src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkBatchInsertRecord.java

```diff
@@ -0,0 +1,322 @@
+                // Strip off 0x if present.
+                String binData = data.toString().trim();
+                if (binData.startsWith("0x") || binData.startsWith("0X")) {
+                    return binData.substring(2);
+                }
+                else {
+                    return binData;
+                }
+            }
+
+            case 2013:    // java.sql.Types.TIME_WITH_TIMEZONE
+            {
+                DriverJDBCVersion.checkSupportsJDBC42();
+                OffsetTime offsetTimeValue;
+
+                // The per-column DateTimeFormatter gets priority.
+                if (null != cm.dateTimeFormatter)
+                    offsetTimeValue = OffsetTime.parse(data.toString(), cm.dateTimeFormatter);
+                else if (timeFormatter != null)
+                    offsetTimeValue = OffsetTime.parse(data.toString(), timeFormatter);
+                else
+                    offsetTimeValue = OffsetTime.parse(data.toString());
+
+                return offsetTimeValue;
+            }
+
+            case 2014: // java.sql.Types.TIMESTAMP_WITH_TIMEZONE
+            {
+                DriverJDBCVersion.checkSupportsJDBC42();
+                OffsetDateTime offsetDateTimeValue;
+
+                // The per-column DateTimeFormatter gets priority.
+                if (null != cm.dateTimeFormatter)
+                    offsetDateTimeValue = OffsetDateTime.parse(data.toString(), cm.dateTimeFormatter);
+                else if (dateTimeFormatter != null)
+                    offsetDateTimeValue = OffsetDateTime.parse(data.toString(), dateTimeFormatter);
+                else
+                    offsetDateTimeValue = OffsetDateTime.parse(data.toString());
+
+                return offsetDateTimeValue;
+            }
+
+            case Types.NULL: {
+                return null;
+            }
+
+            case Types.DATE:
+            case Types.CHAR:
+            case Types.NCHAR:
+            case Types.VARCHAR:
+            case Types.NVARCHAR:
+            case Types.LONGVARCHAR:
+            case Types.LONGNVARCHAR:
+            case Types.CLOB:
+            default: {
+                // The string is copied as is.
+                return data;
+            }
+        }
+    }
+
+    private String removeSingleQuote(String s) {
+        int len = s.length();
+        return (s.charAt(0) == '\'' && s.charAt(len - 1) == '\'') ? s.substring(1, len - 1) : s;
+    }
+
+    @Override
+    public Object[] getRowData() throws SQLServerException {
+        Object[] data = new Object[columnMetadata.size()];
+        int valueIndex = 0;
+        String valueData;
+        Object rowData;
+        int columnListIndex = 0;
+        
+        // check if the size of the list of values = size of the list of columns (which is optional)
+        if (null != columnList && columnList.size() != valueList.size()) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_CSVDataSchemaMismatch"));
```

Comment:
Should we rename `R_CSVDataSchemaMismatch`?

Reply:
It's not a big deal. I'd rather not change an existing entry in the resource file for this.

Reply:
Why wouldn't  you want to change it?

Reply:
This will affect other parts of the program too, and this isn't the only place where the naming change should occur. I'd rather handle these clean-up issues in a separate PR.

Reply:
Turns out the change was pretty simple. I'll make these changes in the next commit.